### PR TITLE
Shortlink with time Layers

### PR DESCRIPTION
**Pull request by RenataMuellerC2C**

### CODE REVIEW

**Comment by sbrunner** [comment]
For me, we should use the same syntax as in the WMS request to have the right precision in the times...

**Comment by RenataMuellerC2C** [comment]
With the same syntax as in the WMS request, layers with resolution 'month', where it is possible to enter a day in the popup calendar, the day will be lost and set to the fist of the month when the link is opened.

File: contribs/gmf/src/datasource/Manager.js [hunk 1]
```diff
@@ -243,6 +243,21 @@ export class DatasourceManager {
...
   }
 
+  /**
+   * Get a datasource by its name.
+   * @param {string} name The name of the datasource.
+   * @hidden
+   */
+  getDatasource(name) {
+    for (const datasourceId in this.dataSourcesCache_) {
+      if (Object.prototype.hasOwnProperty.call(this.dataSourcesCache_, datasourceId)) {
```
Comment (ger-benjamin) [inline contribs/gmf/src/datasource/Manager.js:253]:
```suggestion
      if (this.dataSourcesCache_.hasOwnProperty(datasourceId) {
```
It is not the same in shorter ?

File: contribs/gmf/src/permalink/Permalink.js [hunk 2]
```diff
@@ -1860,6 +1905,43 @@ PermalinkService.prototype.cleanParams = function (groups) {
...
+ * @param {import('ngeo/layertree/Controller.js').LayertreeController} treeCtrl Controller
+ */
+PermalinkService.prototype.setNodeTime_ = function (treeCtrl) {
+  const time = this.ngeoStateManager_.getInitialStringValue(ParamPrefix.TREE_TIME + treeCtrl.node.name);
+  if (time) {
+    const bounds = time.split('/');
+    const node = /** @type {import('gmf/themes.js').GmfGroup|!import('gmf/themes.js').GmfLayerWMS} */ (treeCtrl.node);
+    node.time.minDefValue = bounds[0];
+    node.time.maxDefValue = bounds[1];
+    //let bla = this.ngeoWMSTime.getTime(bounds[0]);
```
Comment (ger-benjamin) [inline contribs/gmf/src/permalink/Permalink.js:1919]: remove me :-)

File: contribs/gmf/src/permalink/Permalink.js [hunk 3]
```diff
@@ -1860,6 +1905,43 @@ PermalinkService.prototype.cleanParams = function (groups) {
...
+    );
+
+    if (dataSource) {
+      if (!(dataSource instanceof ngeoDatasourceOGC)) {
+        throw new Error('Wrong dataSource type');
+      }
+      dataSource.timeLowerValue = new Date(bounds[0]).getTime();
+      dataSource.timeUpperValue = new Date(bounds[1]).getTime();
+    } else if (treeCtrl.children) {
+      for (let i = 0, ii = treeCtrl.children.length; i < ii; i++) {
```
Comment (ger-benjamin) [inline contribs/gmf/src/permalink/Permalink.js:1932]: Please use modern loop: `treeCtrl.children.ForEach((child) => {...})`

File: contribs/gmf/src/datasource/Manager.js [hunk 4]
```diff
@@ -243,6 +243,21 @@ export class DatasourceManager {
     this.ngeoDataSources_.map = map;
   }
 
+  /**
+   * Get a datasource by its name.
+   * @param {string} name The name of the datasource.
+   * @hidden
+   */
+  getDatasource(name) {
```
Comment (ger-benjamin) [inline contribs/gmf/src/datasource/Manager.js:251]: It is secure ? can't we work with ids ? It is possible that we have two datasources with the same name ?

File: contribs/gmf/src/layertree/component.js [hunk 5]
```diff
@@ -525,6 +525,20 @@ Controller.prototype.getNodeState = function (treeCtrl) {
   return treeCtrl.getState();
 };
 
+/**
+ * Update the `timeRangeValue` property of the data source bound to the
+ * given tree controller and the state of the permalink using the given time.
+ *
+ * LayertreeController.prototype.updateTimeData - description
+ * @param {import("ngeo/layertree/Controller.js").LayertreeController} layertreeCtrl ngeo layertree controller
+ * @param {{start : number, end : number}} time The start
```
Comment (ger-benjamin) [inline contribs/gmf/src/layertree/component.js:534]:
```suggestion
 * @param {import("ngeo/datasource/OGC.js").TimeRange} time The start
```
It should works. Please replace also the other definitions.

File: contribs/gmf/src/datasource/Manager.js [hunk 6]
```diff
@@ -243,6 +243,15 @@ export class DatasourceManager {
     this.ngeoDataSources_.map = map;
   }
 
+  /**
+   * Get a datasource by its name.
```
Comment (ger-benjamin) [inline contribs/gmf/src/datasource/Manager.js:247]:
```suggestion
   * Get a datasource by its id.
```

**Review by ger-benjamin — APPROVED** [review]
Looks good, thanks

**Review by sbrunner — APPROVED** [review]

**Pull request merged**